oeps/architectural-decisions/oep-0065-frontend-composability.rst

```diff
@@ -0,0 +1,492 @@
+   * - Status
+     - Under Review
+   * - Type
+     - Architecture
+   * - Created
+     - 2024-04-03
+   * - Review Period
+     - April 15, 2024 - April 29, 2024
+   * - Resolution
+     - <links to any discussions where the final status was decided>
+   * - References
+     -
+       * `FC-0054 - Composable Micro-frontends Discovery <https://openedx.atlassian.net/wiki/spaces/COMM/pages/4063821827/FC-0054+-+Composable+Micro-frontends+Piral+Discovery>`_
+       * `FC-0007 - Modular MFE Domains Discovery <https://openedx.atlassian.net/wiki/spaces/COMM/pages/3614900241/CLOSED+FC-0007+-+Modular+MFE+Domains+Discovery>`_
+
+
+Abstract
+********
+
+This OEP proposes that Open edX micro-frontends (MFEs) adopt module federation as an approach to address inconsistencies, performance issues, and inflexibilities inherent in the current MFE architecture.
+
+Motivation
+**********
+
+Micro-frontends were originally designed to avoid some of the limitations of the monolithic ``edx-platform`` frontend, namely that otherwise independent teams were beholden to the build, test, and release lifecycle of the rest of the codebase. This dramatically slowed down the pace of frontend feature development, experimentation, and innovation.
+
+As a result, the Open edX "micro-frontend" architecture focused on creating an unopinionated, loosely coupled set of applications with the goal of enabling teams to iterate quickly and independently. This goal was successful.
+
+However, now roughly 6 years into this gradual re-platforming, we've discovered that the completely siloed nature of these MFEs has created its own set of problems.  Our frontends are more like independent single-page applications than micro-frontends, as we never invested in ways of composing them together. The problems inherent in today's MFE architecture are described below.
+
+Consistency
+===========
+
+Because each MFE is completely independent, this can create inconsistencies from MFE to MFE.
+
+Reusable components (Paragon)
+-----------------------------
+
+MFEs may use different versions of ``@openedx/paragon``, resulting in functional and stylistic differences.
+
+Header/Footer
+-------------
+
+MFEs may use different versions of ``@edx/frontend-component-header`` and ``@edx/frontend-component-footer``, also resulting in functional, stylistic, and content/navigation differences.
+
+Branding
+--------
+
+Brand packages created from ``@openedx/brand-openedx`` may be different versions, resulting in any number of subtle visual differences.
+
+Other Dependencies
+------------------
+
+MFEs may have completely different versions of any other dependency. We mitigate some of this by consolidating some important dependencies in ``@openedx/frontend-build`` and ``@edx/frontend-platform``, but even those can have different versions from MFE to MFE. For developers, this increases cognitive load and slows velocity because of the need to adjust to the idiosyncracies of each application.
```

Comment:
`@edx/frontend-platform` hasn't migrated to `@openedx`?

Reply:
In general I'm a bit wary of coding in very specific dependency package names within the OEP. What if we change those names in the future, or deprecate a specific package? What if we move the `component-*` packages under `@openedx`? Will we remember, in these cases, to update the OEP?

Reply:
We could just say `frontend-build` and `frontend-platform` for the purposes of this OEP, and I think it's clear what those are.

Reply:
One somewhat-future proof thing we could do would be to have the text link to something (a repo, an npm package?) that will auto-forward if we do change the name. I am a frontend baby so I don't know if npm would do the right thing

Reply:
The way we've handled this on npm has been by updating the README. For example: https://www.npmjs.com/package/@edx/paragon still exists, but when you go there you see
> ## Warning: Package Moved
> While the Paragon design system will continue to receive updates, these updates will no longer be published at the @edx/paragon NPM package. Instead, the latest versions can be found with @openedx/paragon instead.

Reply:
... Yeah, frontend-platform is still `@edx`.  This fact bugged me.  

I'm less concerned with us changing the naming scheme of our frontend repos; how often do we actually rename packages that have been around a while?  Optimizing for that feels excessive, IMHO. 

But maybe I should remove the org prefix from everything and make them links to the repository.  x_x

Reply:
I think I agree that "optimizing for that feels excessive", but I like the idea of making the name a link to npm or github.